File: Codes/2931-maximum-spending-after-buying-items.py
'''
Author: LetMeFly
Date: 2024-12-12 14:37:52
LastEditors: LetMeFly.xyz
LastEditTime: 2024-12-12 14:41:16
'''
from typing import List

class Solution:
    def maxSpending(self, values: List[List[int]]) -> int:
        loc = [len(values[0]) - 1] * len(values)
        ans, cnt, th = 0, len(values) * len(values[0]), 1
        while th <= cnt:
            m, mLoc = 10000000, 0
            for i in range(len(values)):
                if loc[i] >= 0 and values[i][loc[i]] < m:
                    m, mLoc = values[i][loc[i]], i
            ans += th * values[mLoc][loc[mLoc]]
            th += 1
            loc[mLoc] -= 1
        return ans

File: Codes/2931-maximum-spending-after-buying-items.cpp
/*
 * @Author: LetMeFly
 * @Date: 2024-12-12 14:30:55
 * @LastEditors: LetMeFly.xyz
 * @LastEditTime: 2024-12-12 14:36:57
 * AC,100.00%,95.40%
 */
#ifdef _WIN32
#include "_[1,2]toVector.h"
#endif

typedef long long ll;
class Solution {
public:
    ll maxSpending(vector<vector<int>>& values) {
        vector<int> loc(values.size(), values[0].size() - 1);
        ll ans = 0;
        int cnt = values.size() * values[0].size();
        ll th = 1;
        while (cnt--) {
            int m = 1e7, mLoc;
            for (int i = 0; i < values.size(); i++) {
                if (loc[i] >= 0 && values[i][loc[i]] < m) {
                    m = values[i][loc[i]];
                    mLoc = i;
                }
            }
            ans += values[mLoc][loc[mLoc]] * th;
            th++, loc[mLoc]--;
        }
        return ans;
    }
};